sbt 2.x (sbt in Scala 3)

eed3si9n opened the pull request:
This is a work-in-progress port to Scala 3, some of which documented as sudori:

- https://eed3si9n.com/sudori-part1/
- https://eed3si9n.com/sudori-part2/
- https://eed3si9n.com/sudori-part3/

eed3si9n committed:
Cats-like typeclasses

eed3si9n committed:
Reimplement arity-generic list for tuples

eed3si9n committed:
Port Converted

eed3si9n committed:
Port Applicative-do macro to Scala 3

See https://eed3si9n.com/sudori-part3/ for details

eed3si9n committed:
Port StringTypeTag

eed3si9n committed:
Port util-logging

eed3si9n committed:
Port util-complete

eed3si9n committed:
Port util-logic

eed3si9n committed:
Port util-cache

eed3si9n committed:
Port Task

eed3si9n committed:
Refactor AList to handle K[F[a]]

We do need to treat AList like a typeclass after all because
sbt.std.Transform.uniform needs to traverse over a List,
and we won't be able to magically connvert a List into a tuple.

eed3si9n committed:
AList for List

eed3si9n committed:
Port Task Std

eed3si9n committed:
Update TypeFunctions.scala

eed3si9n committed:
sjson-new 0.10.0

eed3si9n committed:
Port testing

eed3si9n committed:
Port run

eed3si9n committed:
Flip to exclude list

eed3si9n committed:
Ignore Scaladoc wrap

eed3si9n committed:
Contraband 0.5.3

eed3si9n committed:
Port commandProj

eed3si9n committed:
Port zinc-lm-integration

eed3si9n committed:
Port sbt client

eed3si9n committed:
Format main-actions

eed3si9n committed:
Format main

eed3si9n commented:
Down to the following subprojects:

```
      actionsProj,
      mainSettingsProj,
      scriptedSbtReduxProj,
      scriptedSbtOldProj,
      scriptedPluginProj,
      dependencyTreeProj,
      mainProj,
      sbtProj,
      bundledLauncherProj,
```

If we ignore the plugin and deployment modules, we're sort of down to the following:

```
      actionsProj,
      mainSettingsProj,
      mainProj,
```

The blockers are:
- [x] Port `Eval.scala` and general `build.sbt` compilation to Scala 3
- [x] Update main macros to Scala 3
- [x] Temporarily fork lm-coursier-shaded against LM 2.0 API (for now same as LM 1.0 but Scala 3)

eed3si9n committed:
Port main-actions

eed3si9n committed:
Port Settings

eed3si9n committed:
Update Scalafmt

eed3si9n committed:
KeyMacro

eed3si9n committed:
InputWrapper can be inline method

eed3si9n committed:
Fix Cont

eed3si9n committed:
Setting macro

eed3si9n committed:
transform can be an inline

eed3si9n committed:
Append macros

eed3si9n committed:
Inline append macros

eed3si9n committed:
Starting input task

eed3si9n committed:
Bump to Scala 3.1.3

eed3si9n committed:
Split buildfile parsing to a subproject

eed3si9n committed:
Manually inline extension method

eed3si9n committed:
Fix conditional task

eed3si9n committed:
Fix Convert

eed3si9n committed:
Fix InputTask macro

eed3si9n committed:
Implement input task helper Def.input

eed3si9n committed:
Refactor to use Type instead of TypeRepr

eed3si9n committed:
valStart to change synthetic variable names

eed3si9n committed:
Fix input task macro

eed3si9n committed:
Fix BSP

eed3si9n committed:
Rename tuple syntax to use mapN

eed3si9n committed:
Migrate updateTask via tuple syntax

eed3si9n committed:
Refactor to build mainProj

eed3si9n committed:
Use Type to compare types instead of TypeRepr

eed3si9n committed:
Use flatMapTask instead of taskDyn

eed3si9n committed:
Fix util-collection test

eed3si9n committed:
Fix task-standard tests

eed3si9n committed:
Some minor tweaks to get things moving

eed3si9n committed:
Fix compilation of the intrinsic Scala version

eed3si9n committed:
Adjust to the code changes

eed3si9n committed:
Comment out 2.13 util tests

eed3si9n committed:
Fix project loading

Problem
-------
Given val root, currently both `root` and synthetic root gets loaded.
This might be caused by build.sbt being virtualized, and no longer
matching the build root directory.

Solution
--------
For now, comparing the canonical paths seems to fix the issue.

eed3si9n committed:
Fix lazy val support

Problem
-------
By Typer phase it seems like lazy val is expanded out to
`def x(): Int` and synthetic `val x$lzy1`,
which results in NoSuchMethod `x$lzy1`.

Solution
--------
Assume the naming convention and grab the part before `$`.

eed3si9n committed:
Comment out tests

eed3si9n committed:
Use launcher scripted tests

eed3si9n committed:
Comment out server test

eed3si9n committed:
Zinc 2.0.0-alpha2

eed3si9n committed:
Use ClassTag instead of Manifest

eed3si9n committed:
bump Zinc, which uses current util

eed3si9n committed:
Scalafmt 3.6.0

eed3si9n committed:
Convert to using slash syntax

eed3si9n committed:
inline parameter removes let binding

eed3si9n committed:
Fix taskDyn

eed3si9n committed:
Turn off turbo mode for now

eed3si9n committed:
sjson-new 0.13.0

eed3si9n committed:
Downgrade scala-xml to 2.13

eed3si9n committed:
Use scala-xml_3 instead

eed3si9n committed:
add Conversion from Project => LocalProject

eed3si9n committed:
config macro

eed3si9n committed:
Mark dependency management tests pending

eed3si9n committed:
Fix InputKey

eed3si9n committed:
Add convertion to ClasspathDep

eed3si9n committed:
Fix setting loading

eed3si9n committed:
Fixes scripted tests

eed3si9n committed:
Minor Scala 3 syntax change

eed3si9n committed:
Fix aggregation

eed3si9n committed:
Adjust to changes in develop branch

eed3si9n committed:
Use extension method to encode the syntax

eed3si9n committed:
lm alpha10

eed3si9n committed:
scripted tests

eed3si9n committed:
def % support

eed3si9n committed:
Add more keys to excludeLintKeys

eed3si9n committed:
Move the project scripted tests to project1

eed3si9n committed:
Interpret bare settings are common settings

See https://eed3si9n.com/simplifying-sbt-with-common-settings/

Problem
-------
The behavior of bare settings is confusing in a multi-project build.
This is partly due to the fact that to use `ThisBuild` scoping
the build user needs to be aware of the task implementation,
and know if the task is already defined at project level.

Solution
--------
This changes the interpretation of the baresettings to be common
settings, which works similar to the way `ThisBuild` behaves in sbt 1.x,
but since this would be a simple append at project-level, it should
work for any tasks or settings.

julianpeeters commented on buildfile/src/main/scala/sbt/internal/Eval.scala:
```diff
@@ -0,0 +1,450 @@
+            if isTopLevelModule(tree.symbol.owner) && isAcceptableType(tpt.tpe) =>
+          val str = name.mangledString
+          vals ::= (
+            if str.contains("$lzy") then str.take(str.indexOf("$"))
```
In case literal identifiers are still supported in 2.x

For example, if I use `sbt new scala/hello-world.g8`:
 - If I first bump `build.properties` to use this 2.x SNAPSHOT
 - and if I then add the following to `build.sbt`: ```lazy val `hello-world` = (project in file("."))```

I then see an error: `java.lang.NoSuchMethodException: $Wrap65ae8c868c$.hello()`, and, digging a little, the val name gets mangled into `public sbt.Project $Wrap65ae8c868c$.hello$minusworld()` where a `$` char has been introduced.
```suggestion
            if str.contains("$lzy") then str.reverse.dropWhile(_ != '$').drop(1).reverse
```

eed3si9n replied:
Thanks!

julianpeeters commented on main/src/main/scala/sbt/internal/GlobalPlugin.scala:
```diff
@@ -20,6 +20,7 @@ import sbt.internal.util.Attributed
 import Def.{ ScopedKey, Setting }
 import Keys._
 import Configurations.{ Compile, Runtime }
+import sbt.ProjectExtra.{ extract, runUnloadHooks, setProject }
```
With the changes to `LoadBuildConfiguration`, it looks like there may need to be an update to the `def build` method, or else global plugins fail to load, due to being passed the wrong `BASE` directory:
```
java.lang.RuntimeException: /Users/julianpeeters/.sbt/2.0.0-alpha6-SNAPSHOT/plugins/global.sbt cannot be mapped using the root paths
Map(
  BASE -> /Users/julianpeeters/Workspace/hello-global,
  SBT_BOOT -> /Users/julianpeeters/.sbt/boot,
  IVY_HOME -> /Users/julianpeeters/.ivy2,
  JAVA_HOME -> /Users/julianpeeters/.sdkman/candidates/java/17.0.1-oracle
)
```
```scala
  def build(base: File, s: State, config: LoadBuildConfiguration): (BuildStructure, State) = {
    val newConverter =
      MappedFileConverter(config.converter.rootPaths2.toMap.updated("BASE", base.toPath), false)
    val newInject =r
      config.injectSettings.copy(global = config.injectSettings.global ++ globalPluginSettings)
    val globalConfig = config.copy(
      injectSettings = newInject,
      pluginManagement = config.pluginManagement.forGlobalPlugin,
      converter = newConverter
    )
```

eed3si9n committed:
Formatting

eed3si9n committed:
Subsume platform cross building

Problem
-------
In sbt 1, platform cross building is implemented using in the user-land
using `%%%` operator, which clevery handles both Scala cross building
and appending platform suffix like sjs1.
However, in general symbolic `%%%` is confusing, and hard to explain.

Solution
--------
In sbt 2, we should subsume the idea of platform cross building,
so `%%` can act as the current `%%%` operator.
This adds a new setting called `platform`, which defaults to
`Platform.jvm` by default.
When a subprojects sets it to `Platform.sjs1`, `ModuleID`s defined using
`%%` operator will inject the platform suffix `_sjs1` **in addition**
to the Scala binary suffix `_2.13` etc.

Note: Explicit JVM dependencies will now require `.platform(Platform.jvm)`.

adpi2 commented on util-collection/src/main/scala-2/sbt/internal/util/HList.scala:
I suggest we remove those Scala 2 files if we don't need to cross-compile `collectionProj` anymore.

adpi2 commented on internal/util-logging/src/main/scala/sbt/util/LogExchange.scala:
```diff
@@ -45,9 +47,11 @@ sealed abstract class LogExchange {
   // Construct these StringTypeTags manually, because they're used at the very startup of sbt
   // and we'll try not to initialize the universe by using the StringTypeTag.apply that requires a TypeTag
   // A better long-term solution could be to make StringTypeTag.apply a macro.
-  lazy val stringTypeTagThrowable = StringTypeTag[Throwable]("scala.Throwable")
-  lazy val stringTypeTagTraceEvent = StringTypeTag[TraceEvent]("sbt.internal.util.TraceEvent")
-  lazy val stringTypeTagSuccessEvent = StringTypeTag[SuccessEvent]("sbt.internal.util.SuccessEvent")
+  lazy val stringTypeTagThrowable = StringTypeTag.manually[Throwable]("java.lang.Throwable")
+  lazy val stringTypeTagTraceEvent =
+    StringTypeTag.manually[TraceEvent]("sbt.internal.util.TraceEvent")
+  lazy val stringTypeTagSuccessEvent =
+    StringTypeTag.manually[SuccessEvent]("sbt.internal.util.SuccessEvent")
```
Since `StringTypeTag.apply` is a macro, we don't need to initialize those manually anymore.

adpi2 commented on util-collection/src/main/scala/sbt/internal/util/Util.scala:
```diff
@@ -45,7 +45,8 @@ object Util {
   def quoteIfKeyword(s: String): String = if (ScalaKeywords.values(s)) s"`${s}`" else s
 
-  def ignoreResult[T](f: => T): Unit = macro Macro.ignore
+  def ignoreResult[A](f: => A): Unit =
+    f; ()
```
Why was it a macro and it is not anymore? Should we inline it?

adpi2 commented on internal/util-complete/src/main/scala/sbt/internal/util/LineReader.scala:
```diff
@@ -236,31 +238,23 @@ abstract class JLine extends LineReader {
 
 @deprecated("Use LineReader apis", "1.4.0")
 private[sbt] object JLine {
-  @deprecated("For binary compatibility only", "1.4.0")
-  protected[this] val originalIn = new FileInputStream(FileDescriptor.in)
-
```
Should we remove `JLine` completely? It only contains deprecated methods.

adpi2 commented on util-collection/src/main/scala/sbt/internal/util/AList.scala:
```diff
@@ -0,0 +1,220 @@
+/**
+ * Arity-generic List. An abstraction over structured Tuple/List type constructor `X1[f[a]]`.
+ */
+trait AList[K[F[_]]]:
```
`AList` on tuples was much more simpler to understand.

> We do need to treat AList like a typeclass after all because
> sbt.std.Transform.uniform needs to traverse over a List,
> and we won't be able to magically connvert a List into a tuple.

We can use `Tuple.fromArray(list.toArray)` instead, if it doesn't impact the performance too much.

adpi2 commented on main-settings/src/main/scala/sbt/Structure.scala:
```diff
@@ -7,187 +7,212 @@
+  final def scopedKey: ScopedKey[A1] = ScopedKey(scope, key)
 
   // @deprecated(Scope.inIsDeprecated, "1.5.0")
-  final def in(scope: Scope): SettingKey[T] =
+  final def in(scope: Scope): SettingKey[A1] =
```
Don't we want to remove this deprecated method? (also the one line 236)

adpi2 commented on main-settings/src/test/scala/sbt/AppendSpec.scala:
```diff
@@ -7,6 +7,7 @@
 
 package sbt
 
+/*
 object AppendSpec {
```
Should we try to uncomment/fix those tests?

(also BuildSettingsInstances, ScopeDisplaySpec, ScopedSpec, SlashSyntaxSpec, SlashSyntaxTest, TaskConfigSpec, TaskPosSpec, TestUtil, TaskNegSpec)

adpi2 commented on buildfile/src/test/scala/sbt/internal/parser/EmbeddedXmlSpec.scala:
```diff
@@ -9,6 +9,7 @@ package sbt.internal.parser
 
 import sbt.internal.util.MessageOnlyException
 
+/*
 object EmbeddedXmlSpec extends CheckIfParsedSpec {
```
Should we uncomment and fix it or remove it? (also SessionSettingsSpec)

adpi2 commented on project/Dependencies.scala:
```diff
@@ -77,16 +79,24 @@ object Dependencies {
   def addSbtZincCompileCore = addSbtModule(sbtZincPath, "zincCompileCore", zincCompileCore)
 
-  val lmCoursierShaded = "io.get-coursier" %% "lm-coursier-shaded" % "2.0.13"
+  // val lmCoursierShaded = "io.get-coursier" %% "lm-coursier-shaded" % "2.0.10"
+  val lmCoursierShaded = "org.scala-sbt" %% "librarymanagement-coursier" % "2.0.0-alpha5"
```
Should we publish the official `lm-coursier-shaded` to Scala 3?

adpi2 commented on main-settings/src/main/scala/sbt/Def.scala:
```diff
@@ -211,69 +213,77 @@ object Def extends Init[Scope] with TaskMacroExtra with InitializeImplicits {
-  def setting[T](t: T): Def.Initialize[T] = macro settingMacroImpl[T]
-  def settingDyn[T](t: Def.Initialize[T]): Def.Initialize[T] = macro settingDynMacroImpl[T]
-  def inputTask[T](t: T): Def.Initialize[InputTask[T]] = macro inputTaskMacroImpl[T]
-  def inputTaskDyn[T](t: Def.Initialize[Task[T]]): Def.Initialize[InputTask[T]] =
```
`inputTaskDyn` is not implemented? Is there any blocker?

adpi2 commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -630,9 +637,9 @@ object Defaults extends BuildCommon {
     resourceGenerators :== Nil,
-    resourceGenerators += Def.task {
+    resourceGenerators += (Def.task {
       PluginDiscovery.writeDescriptors(discoveredSbtPlugins.value, resourceManaged.value)
-    },
+    }).taskValue,
```
Why do we need to use `taskValue`? It seems the type comparer in `TaskMacro` does not work.

adpi2 commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -3602,92 +3669,173 @@ object Classpaths {
+          Boolean,
+          CompatibilityWarningOptions,
+      )
+    ](
+      dependencyResolution,
+      streams,
+      updateConfiguration.toTaskable,
+      (update / logLevel).?.toTaskable,
+      updateCacheName.toTaskable,
+      state,
+      scalaVersion.toTaskable,
+      appConfiguration.toTaskable,
+      Defaults.unmanagedScalaInstanceOnly.toTaskable,
+      dependencyCacheDirectory.toTaskable,
+      crossTarget.toTaskable,
+      executionRoots.toTaskable,
+      resolvedScoped.toTaskable,
+      forceUpdatePeriod.toTaskable,
+      sbtPlugin.toTaskable,
+      thisProjectRef.toTaskable,
+      ivyModule.toTaskable,
+      scalaOrganization.toTaskable,
+      (update / skip).toTaskable,
+      transitiveUpdate.toTaskable,
+      (update / unresolvedWarningConfiguration).toTaskable,
+      evictionErrorLevel.toTaskable,
+      libraryDependencySchemes.toTaskable,
+      assumedEvictionErrorLevel.toTaskable,
+      assumedVersionScheme.toTaskable,
+      assumedVersionSchemeJava.toTaskable,
+      publishMavenStyle.toTaskable,
+      compatibilityWarningOptions.toTaskable,
+    ).mapN {
```
Why do you use the tuple syntax here. Does it not work with `Def.task`? (Same in LibraryManagement.scala:247)

adpi2 commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -1026,7 +1045,8 @@ object Defaults extends BuildCommon {
     cleanFiles := cleanFilesTask.value,
     cleanKeepFiles := Vector.empty,
     cleanKeepGlobs ++= historyPath.value.map(_.toGlob).toVector,
-    clean := Def.taskDyn(Clean.task(resolvedScoped.value.scope, full = true)).value,
+    // clean := Def.taskDyn(Clean.task(resolvedScoped.value.scope, full = true)).value,
```
Any reason not to uncomment this?

adpi2 commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -1076,48 +1096,55 @@ object Defaults extends BuildCommon {
+                  (jar.getName == "scala-library.jar") || (jar.getName.startsWith(
+                    "scala3-library_3"
+                  ))
+                }
+              (allJars.filter { jar =>
+                jar.getName == "scala-compiler.jar" || jar.getName.startsWith("scala3-compiler_3")
+              }) match
+                case Array(compilerJar) if libraryJars.nonEmpty =>
+                  makeScalaInstance(
+                    sv,
+                    libraryJars,
+                    allJars.toSeq,
+                    Seq.empty,
+                    state.value,
+                    scalaInstanceTopLoader.value
+                  )
+                case _ => ScalaInstance(sv, scalaProvider)
             }
-          } else
-          scalaInstanceFromUpdate
+          else scalaInstanceFromUpdate
     }
-  }
 
   // Returns the ScalaInstance only if it was not constructed via `update`
   //  This is necessary to prevent cycles between `update` and `scalaInstance`
   private[sbt] def unmanagedScalaInstanceOnly: Initialize[Task[Option[ScalaInstance]]] =
-    Def.taskDyn {
-      if (scalaHome.value.isDefined) Def.task(Some(scalaInstance.value)) else Def.task(None)
+    (Def.task { scalaHome.value }).flatMapTask { case h =>
```
It seems that you managed to port `Def.taskDyn`. So I would prefer to revert those changes, if that's okay.

adpi2 commented on main/src/main/scala/sbt/internal/AddSettings.scala:
```diff
@@ -23,15 +24,15 @@ object AddSettings {
-    override def toString: String = s"SbtFiles($files)"
-  }
+  private[sbt] final class DefaultSbtFiles(val include: VirtualFile => Boolean) extends AddSettings
+  // private[sbt] final class SbtFiles(val files: Seq[File]) extends AddSettings {
```
What are `SbtFiles` and what does it imply to remove them?

adpi2 commented on main/src/main/scala/sbt/internal/FileChangesMacro.scala:
```diff
@@ -14,91 +14,91 @@ import sbt.nio.Keys._
     )
-    def inputFileChanges: FileChanges = macro changedInputFilesImpl[T]
+    inline def inputFileChanges: FileChanges =
+      ${ FileChangesMacro.changedInputFilesImpl[A]('in) }
```
Does it make sense to have a `@compileTimeOnly` on a macro def. It should always be expanded.

So I think we could remove the annotations.

adpi2 commented on main/src/test/scala/PluginCommandTest.scala:
```diff
@@ -7,9 +7,11 @@
 
 package sbt
 
+/*
```
Should we try to uncomment/fix those tests or remove them? (also PluginTests, DefinitionTest and SettingQueryTest)

adpi2 commented on sbt-app/src/test/scala/sbt/IllegalReferenceSpec.scala:
```diff
@@ -10,6 +10,7 @@ package sbt
 import org.scalatest
 import org.scalatest.{ TestData, fixture, funsuite }
 
+/*
```
Same here. Fix or remove?

adpi2 commented on project/Scripted.scala:
```diff
@@ -146,15 +158,26 @@ object Scripted {
           callback,
-          scalaVersion,
-          sbtVersion,
-          classpath.toArray,
```
Do you remember the purpose of these changes?

adpi2 commented on build.sbt:
```diff
@@ -10,14 +10,14 @@ import scala.util.Try
 ThisBuild / scalafmtOnCompile := !(Global / insideCI).value
 ThisBuild / Test / scalafmtOnCompile := !(Global / insideCI).value
-ThisBuild / turbo := true
+// ThisBuild / turbo := true
```
Why disabling turbo mode? Does it make the tests flaky?

adpi2 commented on main/src/main/scala/sbt/internal/LintUnused.scala:
```diff
@@ -31,17 +32,24 @@ object LintUnused {
       onLoad,
       onLoadMessage,
       onUnload,
+      pollInterval,
+      pushRemoteCacheArtifact,
+      sbt.nio.Keys.outputFileStamper,
       sbt.nio.Keys.watchTriggers,
       serverConnectionType,
       serverIdleTimeout,
       shellPrompt,
+      sLog,
+      traceLevel,
```
Do you have a clue why we need to add more keys to `excludeLintKeys`?

adpi2 commented on sbt-app/src/sbt-test/project1/Class.forName/build.sbt:
Do you remember why you renamed `project` to `project1`?

adpi2 commented on core-macros/src/main/scala-2/sbt/internal/util/appmacro/LinterDSL.scala:
LinterDS and TaskLinterDSL are not ported? What are they supposed to do? Should we try to port them or remove them?

adpi2 reviewed (approved):
Thanks Eugene, that's huge!

I took the time to thoroughly review it and I now have a much better understanding of what it takes to port sbt to Scala 3, and what we still need to do.

I'm on board with merging it now and continue working on the migration in separate PRs.

## General Questions

I would appreciate some answers if you have time to think about the following questions.

1. Many projects are configured with cross-compilation and many of them don't cross compile anymore. Do you know if there are some projects that we should keep cross-compiling? Or if we should drop cross-compilation in this branch. #7433
2. Should we avoid using the new syntax to make it smoother to port back and from sbt 1.x

You will also find more questions in the review comments.
 
## TODO (mostly for myself)

Below is a list of ideas I noted during the review. After we merge this PR, I'll open new ones to propose changes.

<details>
<summary>TODO (mostly for myself)</summary>

- [ ] Update to Scala 3.3.1
- [ ] Remove cross-compilation options from `utilCommonSettings` #7433
- [ ] Fix or remove cross-compilation in `testingProj`, `utilScripted`, `protocolProj`, `stdTaskProj`, `logicProj`, `zincLmIntegrationProj`, `taskProj`, `sbtClientProj`, `runProj`, `mainProj`, `buildFileProj`, `completeProj`, `coreMacrosProj`, `mainSettingsProj`, `utilLogging`, `utilCache`, `actionsProj`, `collectionProj`, `commandProj`, `utilTracking` (it still works in `utilControl`, `bundledLauncherProj`, `scriptedPluginProj`, `utilRelation`, `utilPosition`) #7433
- [ ] Try declare `F` as covariant in `Execute`, `ExecuteProgress` and `PMap` #7442
- [ ] Try get rid of the obscure `~>|` in `TypeFunctions`
- [ ] Can we drop the support of List in `AList`. (we can convert `List` to `Tuple` using `Tuple.fromArray(list.toArray)`)
- [ ] `*` in `RichOps` is hiding `*` in `StringOps`, find a fix (an overload?)
- [ ] In Cont.scala, use `report.errorAndAbort` instead of `sys.error`
- [ ] Try define `RichTaskable` as `RichTaskable[T <: Tuple](t: Tuple.Map[T, ST])`
- [ ] In Structure.scala, get rid of `Manifest`, maybe introduce `TaskTag[A]`, `InputTaskTag[A]`...
- [ ] List.scala:115-147, define `TupleK` as `type TupleK[T <: Tuple] = [F[_]] =>> Tuple.Map[T, F]`
- [ ] In TaskMacro.scala lines 108, 114, 122...: inline the report.errorAndAbort
- [ ] Fix todo in ErrorSpec.scala:38, and try uncomment test (line 43)
- [ ] Dependencies.scala line 94: publish eval for Scala 3.3.1
- [ ] Fix todo in SbtRefactorings.scala:96
- [ ] Fix todo in Main.scala:134-135: store `bootServerSocket` in state
- [ ] Fix todo in Load.scala:746 (BSP reporter)
- [ ] Fix todo in Load.scala:280
- [ ] Fix todo in Opts.scala:47&53 (Sonatype OSS repos)
- [ ] Fix todo in Import.scala:301-302
- [ ] Uncomment and fix tests in ci.yml (lines 115, 121, 122, 126, 127, 129, 130, 136, 141, 143, 144, 149, 151)
- [ ] Uncomment and fix tests: (ParserKey.scala, ParserSpec.scala, PluginCommandTest.scala, CompletionSpec.scala)
- [ ] Fix pending or commented scripted tests:
  `clean-managed`, `compile-clean`, `compile-time-only`, `dependency-management/artifact/build.sbt:14`, `actions/cross-advanced/test:22-27`, `actions/depends-on/build.sbt:5-7`, `input-task-dyn`, `generator`, `run-task`, `task-map`, `src-dep-plugin`, `exclude-dependencies`, `exclude-dependencies2`, `pom-advanced`, `version-interval`, `no-scala-tool`, `auto-plugins-default-requires-jvmplugin`, `binary-plugin`, `derived`, `extra`, `giter-8`, `ignore-hidden-build-files`, `fork-loader`, `non-local-main`, `spawn-exit`, `scala-instance-classloader`, `serial`
- [ ] Port or remove commented Scala 2 code in
  Types.scala, TypeFunctions.scala, Actions.scala, Def.scala (line 310, 318, 421), InputTask.scala, InputWrapper.scala (line 15, 67, 161, 221), , Instances.scala (line 38), TaskMacro.scala, IntputTaskMacro.scala (line 116), main.contra, ProjectOrigin.scala, PluginTrigger.scala, UsageTest.scala, ProjectExtra.scala, StateTransform.scala, Import.scala (line 138, 144, 159-164...)
- [ ] Remove deprecated methods in Structure.scala
- [ ] Make a community build of plugins to enable global-plugins
- [ ] Port all PRs from 1.10.x

</details>

eed3si9n commented:
> Many projects are configured with cross-compilation and many of them don't cross compile anymore. Do you know if there are some projects that we should keep cross-compiling? Or if we should drop cross-compilation in this branch.

We should probably drop Scala 2.x cross build.

> Should we avoid using the new syntax to make it smoother to port back and from sbt 1.x

In some situations maybe, but you may have noticed I am a big fan of the new syntax, so I think we should just use Scala 3 as intended with all the bells and whistles.

eed3si9n commented:
> I'm on board with merging it now and continue working on the migration in separate PRs.

Yea. Let's merge this as-is, and I'll come back to this PR to respond to comments and todos.
